### PR TITLE
`Alternative` for `ParsecT`

### PR DESCRIPTION
**Pull request by DanielRrr**
`Alternative` implementation for `ParsecT` and some module restructurisation

### CODE REVIEW

File: Megaparsec/Megaparsec.lean [hunk 1]
```diff
@@ -3,6 +3,34 @@ import Mathlib.Algebra.Group.Defs -- for Monoid
 
 namespace Megaparsec
 
+/-- Util -/
+inductive Maybe (R: Type u) where
+| just (ok: R)
+| nothing
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:9]: We can just use `Option` instead

File: Megaparsec/Megaparsec.lean [hunk 2]
```diff
@@ -69,34 +104,60 @@ inductive ErrorFancy (E: Type) where
...
+               | (Maybe.nothing, Maybe.nothing) => ParseError.trivial s₁ Maybe.nothing (p₁ ++ p₂)
+               | (Maybe.just x, Maybe.just y) => ParseError.trivial s₁ (Maybe.just (errorItemMax x y)) (p₁ ++ p₂)
+               | (Maybe.nothing, Maybe.just x) => ParseError.trivial s₁ (Maybe.just x) (p₁ ++ p₂)
+               | (Maybe.just x, Maybe.nothing)=> ParseError.trivial s₁ (Maybe.just x) (p₁ ++ p₂)
+          | (ParseError.fancy _ _, ParseError.trivial _ _ _) => e₁
+          | (ParseError.trivial _ _ _, ParseError.fancy _ _) => e₂
+          | (ParseError.fancy s₁ x₁, ParseError.fancy _ x₂) => ParseError.fancy s₁ (x₁ ++ x₂)
+    | Ordering.gt => e₁
+
+def toHints [s : Stream S] (streamPos : ℕ) (e : @ParseError S E s) : Hints (s.Token) :=
```
Comment (arthurpaulino) [inline Megaparsec/Megaparsec.lean:137]:
```suggestion
def toHints [s : Stream S] (streamPos : ℕ) (e : @ParseError S E s) : Hints s.Token := 
```

**Review by arthurpaulino — CHANGES_REQUESTED** [review]

**Comment by arthurpaulino** [comment]
lgtm but let's wait for @cognivore's input

**Review by cognivore — APPROVED** [review]
LGTM!

Option is nice find!

**Pull request merged**